### PR TITLE
ci: added a link-check ci

### LINKED ISSUE #103: Improve the CI/CD pipeline to avoid broken links
After the site roll-out we found some broken links. In order to avoid this in the future we should improve the merge checks:

- [ ] Add workflow to validate broken links.

GA Example to validate links:
https://github.com/openjs-foundation/cross-project-council/actions/runs/6412837418/workflow

### PR DESCRIPTION
**Pull request by utnim2**
Fixes #103 

follow up to #291

### CODE REVIEW

**Comment by Relequestual** [comment]
I think this is blocked by https://github.com/json-schema-org/website/issues/295

If we merged this before actually fixing links, it would block PRs till the bad links are fixed, I think.

**Comment by Min2who** [comment]
> If we merged this before actually fixing links, it would block PRs till the bad links are fixed, I think.

Lets wait for #295 to be solved

**Comment by DarhkVoyd** [comment]
Hey, I am not sure if linkinator is the right tool for checking broken links for our use-case.

**Comment by Min2who** [comment]
> Hey, I am not sure if linkinator is the right tool for checking broken links for our use-case.

reason will be appreciated @DarhkVoyd

**Comment by Min2who** [comment]
Hey @Relequestual can you review it made some changes here

File: .github/workflows/link-check.yml
Comment (DarhkVoyd) [inline .github/workflows/link-check.yml:1]: Yes, the previous link checker had some issues. This one here is the link checker I was referring to use instead. Still, there are a few config tweaks we'll need. I'll drop my suggestions once the PR for fixing the current link issues is good to go.

**Review by DarhkVoyd — CHANGES_REQUESTED** [review]

File: pages/blog/posts/bowtie-intro.md
Comment (Min2who) [reply pages/blog/posts/bowtie-intro.md:1]: > additional changes alongside the pull request might not be the best practice. Additionally, there exists a separate pull request proposing this and more link changes.
what additional changes are you refering to ?

Comment (DarhkVoyd) [reply pages/blog/posts/bowtie-intro.md:1]: I was referring to changing / fixing the blog link.

Comment (Min2who) [reply pages/blog/posts/bowtie-intro.md:1]: > I was referring to changing / fixing the blog link.

It was done to trigger the workflow :)

Comment (DarhkVoyd) [reply pages/blog/posts/bowtie-intro.md:1]: No big deal, I was just point that out. Cause, the change ended up being a part of the PR, committing changes to the PR might not be the way to test.

File: .github/workflows/link-check.yml
Comment (Relequestual) [reply .github/workflows/link-check.yml:1]: Feel free to add suggestions now. If you comment on a specific line or several lines, you can [add your suggestion as part of a comment](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request).

File: .github/workflows/link-check.yml
Comment (Relequestual) [inline .github/workflows/link-check.yml:10]: I don't think we want to limit the link checker to only work for the blog. Should be for the whole site.

File: .lycheeignore
Comment (Relequestual) [reply .lycheeignore:1]: There may be cases in the old specification documents which are valid URIs but not valid URLs, and so they will not be reachable. Having this hear is useful I think.
@Min2who I would add something to the README.md that mentions that links are checked and can be ignored/skipped using this file.

File: pages/blog/posts/bowtie-intro.md
Comment (Relequestual) [reply pages/blog/posts/bowtie-intro.md:1]: > > I was referring to changing / fixing the blog link.
> 
> It was done to trigger the workflow :)

OK. You can always do a follow up commit to revert the changes, or better still, interactive git rebase to remove that commit, then force push. (Although you might want to practice that on another repo first so you are confident you're doing it right. Interactive git rebase is a great tool to learn!)

**Review by Relequestual — CHANGES_REQUESTED** [review]
Moving in the right direction. Made a few comments for change. Thanks for your continued effort on this work! =D

File: .github/workflows/link-check.yml
Comment (DarhkVoyd) [reply .github/workflows/link-check.yml:10]: Here, I would suggest to run this workflow only in case of pushes to the main branch and manual workflow dispatch and not in case of pull requests because of the way lychee will check our relative links i.e. lychee will reach out to the deployed website and attempt to resolve our relative links from the root page of the website which will often fail because the PR changes wouldn't be reflected on the deployment. 

Note: When using relative links in our blogs we will have to use root paths `/blog/posts/hello-world-hello-postman` instead of `./hello-world-hello-postman`.

File: .github/workflows/link-check.yml
Comment (DarhkVoyd) [inline .github/workflows/link-check.yml:22]: Next, here we would have to set base to `--base https://json-schema.org/` so that all the relative links are resolved from the base of our deployment.

File: .github/workflows/link-check.yml
Comment (DarhkVoyd) [reply .github/workflows/link-check.yml:1]: @Relequestual I have added some suggestions, kindly have a look. Lastly, there is a flag in Lychee that automates the creation of an issue each time it finds a broken link; using it is your call.

File: .github/workflows/link-check.yml
Comment (Relequestual) [reply .github/workflows/link-check.yml:10]: No, we DO want it to be on pull request. We want to avoid deploying with a broken link.
We need to work out how to also check the deployable version in the PR.
It might be a case of running the build before link checking, then doing a remap as part of the config to address localhost.

Comment (DarhkVoyd) [reply .github/workflows/link-check.yml:10]: Absolutely, thank you for your patience on this. I'll work on a solution to check the deployable version.

**Comment by Min2who** [comment]
@Relequestual made some changes a review will be appreciated and left some comment in the  code

File: .lycheeignore
Comment (benjagm) [inline .lycheeignore:1]: Why this file in the root folder? Why not in the same folder than the workflow code?

File: pages/blog/posts/bowtie-intro.md
Comment (benjagm) [inline pages/blog/posts/bowtie-intro.md:48]: Please don'f fix link as part of this PR. Let's keep it just for the code.
Better discard this change.

We have this PR for this: #299

File: .github/workflows/link-check.yml
Comment (benjagm) [reply .github/workflows/link-check.yml]: Sound good!

File: .github/workflows/link-check.yml
Comment (benjagm) [inline .github/workflows/link-check.yml:46]:
```suggestion
          args: '--base https://json-schema.org --verbose --no-progress **/*.md' 
```

Comment (benjagm) [inline .github/workflows/link-check.yml:46]: **/*.md 

What is the meaning of this parameter?

**Review by benjagm — CHANGES_REQUESTED** [review]
We very close. Nice job!

Left some comment.

File: pages/blog/posts/bowtie-intro.md
Comment (benjagm) [inline pages/blog/posts/bowtie-intro.md:48]: @utnim2 Lets remove this change from the PR. Let's keep just the workflow.

**Review by benjagm — APPROVED** [review]
I havent tested it but looks good to me.

**Comment by benjagm** [comment]
@aialok can you please join the discussion to help us with a PR review of the great work done here by @utnim2 ?

Alok you are working in other CI/CD issue so I think your review here will be very valuable.

**Comment by aialok** [comment]
> @aialok can you please join the discussion to help us with a PR review of the great work done here by @utnim2 ?
> 
> Alok you are working in other CI/CD issue so I think your review here will be very valuable.

Sure @benjagm : )

File: .github/workflows/link-check.yml
Comment (aialok) [inline .github/workflows/link-check.yml:43]: If you always want to use the latest features and avoid breaking changes, you can replace the version with `lycheeverse/lychee-action@v1`.

Reference: https://github.com/lycheeverse/lychee-action - checkout README.md Usage Section

File: .github/workflows/link-check.yml
Comment (utnim2) [reply .github/workflows/link-check.yml:46]: all the markdown file

File: .github/workflows/link-check.yml
Comment (DarhkVoyd) [inline .github/workflows/link-check.yml:46]: Hey, 
To address @Relequestual 's feedback, we should revert my suggestion of setting base as the deployed build. As, deployed build with not include changes introduced in a PR.

> No, we DO want it to be on pull request. We want to avoid deploying with a broken link.
We need to work out how to also check the deployable version in the PR.
It might be a case of running the build before link checking, then doing a remap as part of the config to address localhost.

**Comment by aialok** [comment]
@utnim2, please fix your build failures so that we can determine whether the linkchecker steps are working correctly or not. To address the build issues, you can check out the following workflow: 
https://github.com/Lakhua-dev/website/actions/runs/8128474680/job/22214405519

Screenshot from 2024-03-03 12-48-23

Basically, it is failing because the markdown file in above image, i.e., `_index.md`, in both directories `pages/overview/code-of-conduct` and `pages/overview/sponsors`, is located in a git submodule. Therefore, you need to import the git submodule as well.

Thank You : )
Happy coding!

File: .github/workflows/link-check.yml
Comment (aialok) [inline .github/workflows/link-check.yml:25]: Use `yarn install --frozen-lockfile` instead.

You can check out this npm blog: [Introducing npm ci for faster, more reliable builds](https://blog.npmjs.org/post/171556855892/introducing-npm-ci-for-faster-more-reliable).

`yarn install --frozen-lockfile` does the same thing as `npm ci`. Since we are using Yarn, we are good to go.

File: .github/workflows/link-check.yml
Comment (aialok) [inline .github/workflows/link-check.yml:40]: You should add step names for maintaining consistency in the workflow.
```yaml
- name: Checkout Repository

**Comment by aialok** [comment]
Thank you for your changes, @utnim2. To ensure that the `Build Job` doesn't fail, we need to include an additional line to fetch submodules during the repository checkout process. Adding the following line will resolve the issue:

```yaml
- name: Checkout repository
  uses: actions/checkout@v4
  with:
    submodules: recursive
```
Checkout this for more details : https://github.com/json-schema-org/website/pull/297#issuecomment-1975072270

File: .github/workflows/link-check.yml
Comment (aialok) [inline .github/workflows/link-check.yml:32]: I think you need to remove this steps and add this to linkchecker job as we need our app to run locally so that linkchecker can do it job.
```suggestion
     Remove this step
```

File: .github/workflows/link-check.yml
Comment (aialok) [inline .github/workflows/link-check.yml:46]: Add steps to run our app locally.

**Review by aialok — COMMENTED** [review]
@utnim2 , See Comment : )

File: .github/workflows/link-check.yml
Comment (utnim2) [reply .github/workflows/link-check.yml:32]: Thanks for your review @aialok 

no we need to first build and deploy it here

The deploy step is missing here as the deployment is not yet clear

@aialok

File: .github/workflows/link-check.yml
Comment (utnim2) [reply .github/workflows/link-check.yml:46]: see comment [here](https://github.com/json-schema-org/website/pull/297#discussion_r1511154012)

File: .github/workflows/link-check.yml
Comment (DarhkVoyd) [reply .github/workflows/link-check.yml:32]: Yes, I too agree with you regarding to need to build and deploy.

Comment (aialok) [reply .github/workflows/link-check.yml:32]: > Thanks for your review @aialok
> 
> no we need to first build and deploy it here and pass the deployed url in the linkchecker action (i have put localhost:3000 as of now)
> 
> The deploy step is missing here as the deployment is not yet clear
> 
> @aialok

Got it man. But I still have one doubt what do you mean by deploy locally. Deploy locally means running on localhost?

Thank you @utnim2. You have done pretty good job : )

**Comment by utnim2** [comment]
> With reference to this comment : https://github.com/json-schema-org/website/pull/297#discussion_r1511154012
> 
> I think we do already deploy our app on cloudflare. 
> We can directly use `https://json-schema.org/` instead of this `http://localhost:3000` as suggested in comment: https://github.com/json-schema-org/website/pull/297#discussion_r1501556123
> 
> @utnim2 @DarhkVoyd

**Comment by DarhkVoyd** [comment]
> With reference to this comment : [#297 (comment)](https://github.com/json-schema-org/website/pull/297#discussion_r1511154012)
> 
> I think we do already deploy our app on cloudflare. We can directly use `https://json-schema.org/` instead of this `http://localhost:3000` as suggested in comment: [#297 (comment)](https://github.com/json-schema-org/website/pull/297#discussion_r1501556123)
> 
> @utnim2 @DarhkVoyd

Please refer to https://github.com/json-schema-org/website/pull/297#discussion_r1507889845

**Comment by aialok** [comment]
> > With reference to this comment : [#297 (comment)](https://github.com/json-schema-org/website/pull/297#discussion_r1511154012)
> > I think we do already deploy our app on cloudflare. We can directly use `https://json-schema.org/` instead of this `http://localhost:3000` as suggested in comment: [#297 (comment)](https://github.com/json-schema-org/website/pull/297#discussion_r1501556123)
> > @utnim2 @DarhkVoyd
> 
> Please refer to [#297 (comment)](https://github.com/json-schema-org/website/pull/297#discussion_r1507889845)

My Bad...Got it : )

**Review by DarhkVoyd — CHANGES_REQUESTED** [review]
Hello @benjagm @utnim2 @aialok 

This PR was quite difficult and posed some challenges during review. So, please bear with me if I make some mistakes and I appreciate your patience as I worked through it. Below are my findings, change requests, and suggestions:

1. I've noticed that we're encountering issues when using the frozen lockfile during dependency changes. We should explore potential solutions to avoid such problems.
2. We should consider either merging the two jobs or rebuilding and serving the website again in the second job to address post-action cleanup in the workflow.
3. Despite the expected downtime of our served website due to post-cleanup actions, the initial job fails to serve the website. The current workflow experiences a silent failure in the first job due to a command not found error when serving the website.
Upon investigation, it appears that the `serve` package, utilized in our `package.json` for local website deployment, is missing from our dependencies. This absence is the root cause of the deployment failure. To resolve this issue, we must ensure the inclusion of the `serve` package in our dependencies.
4. Given that we use `localhost` as the base for our action, we need to define which files should be considered part of the action. Because, I am not sure if all markdown file references are served on our website this ensures we avoid unnecessary errors.
5.  In my testing of the workflow, I found several broken links. To prevent workflow failure, I suggest creating a separate issue to address these broken links before merging this pull request. The broken links may be related to unnecessary files from submodules or other issues discussed in point 4.


Thank you for your patience. For a more detailed overview of my changes, please visit [this link](https://github.com/DarhkVoyd/website/pull/4/files).

**Comment by utnim2** [comment]
@DarhkVoyd thanks for you review
Can you refer [this](https://github.com/json-schema-org/website/pull/297#discussion_r1511154012) comment about the deploy part

**Comment by utnim2** [comment]
Can you add your opinion @Relequestual

**Comment by benjagm** [comment]
> Can you add your opinion @Relequestual

Can we just validate the external links for now and leave the internal ones for a second iteration?
Reference [here](https://github.com/wummel/linkchecker/issues/547)

**Comment by benjagm** [comment]
An alternative approach can be pause this effort to validate broken links in deployment time, to use a different approach to have a crawler finding broken links every X days. For example:

https://github.com/marketplace/actions/broken-links-crawler

**Comment by DarhkVoyd** [comment]
> @DarhkVoyd thanks for you review Can you refer [this](https://github.com/json-schema-org/website/pull/297#discussion_r1511154012) comment about the deploy part

utnim2, 
I believe you have misinterpreted my suggestion.

**Comment by aialok** [comment]
> Hello @benjagm @utnim2 @aialok
> 
> This PR was quite difficult and posed some challenges during review. So, please bear with me if I make some mistakes and I appreciate your patience as I worked through it. Below are my findings, change requests, and suggestions:
> 
> 1. We should consider either merging the two jobs or rebuilding and serving the website again in the second job to address post-action cleanup in the workflow.
> 2. Despite the expected downtime of our served website due to post-cleanup actions, the initial job fails to serve the website. The current workflow experiences a silent failure in the first job due to a command not found error when serving the website.
>    Upon investigation, it appears that the `serve` package, utilized in our `package.json` for local website deployment, is missing from our dependencies. This absence is the root cause of the deployment failure. To resolve this issue, we must ensure the inclusion of the `serve` package in our dependencies.
> 3. Given that we use `localhost` as the base for our action, we need to define which files should be considered part of the action. Because, I am not sure if all markdown file references are served on our website this ensures we avoid unnecessary errors.
> 4. In my testing of the workflow, I found several broken links. To prevent workflow failure, I suggest creating a separate issue to address these broken links before merging this pull request. The broken links may be related to unnecessary files from submodules or other issues discussed in point 4. We might also use `.lycheeignore` to ignore certain URIs or URLs.
> 
> Thank you for your patience. For a more detailed overview of my changes, please visit [this link](https://github.com/DarhkVoyd/website/pull/4/files).

Great review! Thanks, @DarhkVoyd.

I've gone through your workflow, and everything looks good to me. However, I noticed that we are checking links of `.md` files in `node_modules`, which need to be ignored as they are causing thousands of links to be checked unnecessarily.

Screenshot from 2024-03-06 14-43-28

Other than this, everything looks good to me.

@benjagm, your solution is also a good way to handle things.

Happy coding!

**Comment by DarhkVoyd** [comment]
> Great review! Thanks, @DarhkVoyd.
I've gone through your workflow, and everything looks good to me. However, I noticed that we are checking links of .md files in node_modules, which need to be ignored as they are causing thousands of links to be checked unnecessarily.

**aialok**, 
Thank you so much for the review. I too was confused about several unnecessary files coming up in the workflow but could not understand the origination as expressed in my review. 

**benjagm**, 
I agree with your solution and I believe that it is the only correct implementation for this workflow. Since, we want to maintain our relative paths in the markdown for relative links we probably will have to break down our types of links not only into external and internal but into relative paths too.

**Comment by Relequestual** [comment]
> > Can we just validate the external links for now and leave the internal ones for a second iteration? Reference [here](https://github.com/wummel/linkchecker/issues/547)
> 
> I too agree with this

I'm not concerned much about external links, as they can be broken any time. I'm much more concerned about internal links. We should avoid introducing changes which cause internal links to break. That should be our primary objective here (at least to me).

**Review by Relequestual — CHANGES_REQUESTED** [review]
Let's avoid checking the `node_modules` folder =] .


Is it possible to only check internal links?
I don't mind also checking external links, but I don't think it's as needed or the intent of the Issue.

**Comment by utnim2** [comment]
> Let's avoid checking the `node_modules` folder =] .
> 
> Is it possible to only check internal links? I don't mind also checking external links, but I don't think it's as needed or the intent of the Issue.

So i need to modify the action so that it checks only for external links right?

cc @benjagm @Relequestual

**Comment by benjagm** [comment]
> So i need to modify the action so that it checks only for external links right?

What Ben was saying is check only internal but the problem with internal links but as we have been discussing this approach seems to not work for the specific type of architecture and stack of this site.

**Comment by benjagm** [comment]
Closing this as there was not activity for the last weeks.

**Pull request closed without merging**
